clarifai/datasets/upload/loaders/coco_segmentation.py

```diff
@@ -2,165 +2,94 @@
-              annots.append(poly.tolist())  #[[x=col, y=row],...]
-              class_names.append(self.cat_id_map[cat_id])
-          else:  # seg: {"counts":[...]}
-            if isinstance(ann['segmentation']['counts'], list):
-              rle = maskUtils.frPyObjects([ann['segmentation']], image_height, image_width)
-            else:
-              rle = ann['segmentation']
-            mask = maskUtils.decode(rle)  #binary mask
-            #convert mask to polygons and add to annots
-            contours, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
-            polygons = []
-            for cont in contours:
-              if cont.size >= 6:
-                polygons.append(cont.astype(float).flatten().tolist())
-            # store polygons in (x,y) pairs
-            polygons_flattened = reduce(lambda x, y: x + y, polygons)
-            del polygons
-            del contours
-            del mask
-            gc.collect()
-
-            polygons = np.array(polygons_flattened).reshape((int(len(polygons_flattened) / 2), 2))
-            polygons[:, 0] = polygons[:, 0] / image_width
-            polygons[:, 1] = polygons[:, 1] / image_height
-
-            annots.append(polygons.tolist())  #[[x=col, y=row],...,[x=col, y=row]]
-            class_names.append(self.cat_id_map[cat_id])
-      else:  # if no annotations for given image_id-cat_id pair
-        continue
-    assert len(class_names) == len(annots), f"Num classes must match num annotations\
-    for a single image. Found {len(class_names)} classes and {len(annots)} polygons."
-
-    return VisualSegmentationFeatures(image_path, class_names, annots, id=str(_id))
+    concept_ids = []
+
+    input_ann_ids = self.coco.getAnnIds(imgIds=[value['id']])
+    input_anns = self.coco.loadAnns(input_ann_ids)
+
+    for ann in input_anns:
+      # get concept info
+      # note1: concept_name can be human readable
+      # note2: concept_id can only be alphanumeric, up to 32 characters, with no special chars except `-` and `_`
+      concept_name = self.coco.cats[ann['category_id']]['name']
+      concept_id = concept_name.lower().replace(' ', '-')
+
+      # get polygons
+      if isinstance(ann['segmentation'], list):
+        poly = np.array(ann['segmentation']).reshape((int(len(ann['segmentation'][0]) / 2), 2))
+        annots.append(poly.tolist())  #[[x=col, y=row],...]
+        poly[:, 0], poly[:, 1] = poly[:, 0] / value['width'], poly[:, 1] / value['height']
+        concept_ids.append(concept_id)
+      else:  # seg: {"counts":[...]}
+        if isinstance(ann['segmentation']['counts'], list):
+          rle = maskUtils.frPyObjects([ann['segmentation']], value['height'], value['width'])
+        else:
+          rle = ann['segmentation']
+        mask = maskUtils.decode(rle)  #binary mask
+        #convert mask to polygons and add to annots
+        contours, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
```

Comment:
Just found this, and is out of scope of this PR. Annotations support masks as base64 images: Annotation -> Data -> Image. How come we are converting them into polygons?

Reply:
For segmentation, only polygons are supported in dataloader
https://github.com/Clarifai/clarifai-python/blob/master/clarifai/datasets/upload/features.py

Reply:
Why is that, when we support seg masks?

Reply:
This is in todo planning, Need to extend the support for mask uploads.